File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDiameter = 0;
    int diameterOfBinaryTree(TreeNode* root)
    {
        height(root);
        return maxDiameter;
    }
    
    int height(TreeNode* root)
    {
        if(root == nullptr)
        return 0;

        int left = height(root->left);
        int right = height(root->right);
        maxDiameter = max(maxDiameter, left + right);
        return 1 + max(left, right);
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
class Solution {
    int maxDiameter = 0;

    public int checker(TreeNode root) {
        if (root == null) return 0;
        
        int left = checker(root.left);
        int right = checker(root.right);
        
        // Update the global max with the path through the current node
        maxDiameter = Math.max(maxDiameter, left + right);
        
        // Return the height of the current node
        return 1 + Math.max(left, right);
    }

    public int diameterOfBinaryTree(TreeNode root) {
        maxDiameter = 0; // <--- RESET HERE
        checker(root);
        return maxDiameter;
    }
}
